app/codes/state_updater.py

```diff
@@ -106,7 +115,80 @@ def update_state_from_transaction(cur, transaction_type, transaction_data, trans
         )
 
 
-
+def update_state_from_sc_tx(cur, transaction_data, transaction_code, transaction_timestamp,transaction_signer=None):
+    """Handling smart contract transaction differently from the rest"""
+    funct = transaction_data['function']
+    if funct == "setup":  # sc is being set up
+        contract = dict(transaction_data['params'])
+        transaction_data['params']['parent'] = transaction_code
+    else:
+        contract = get_contract_from_address(
+            cur, transaction_data['address'])
+    module = importlib.import_module(
+        ".codes.contracts."+contract['name'], package="app")
+    sc_class = getattr(module, contract['name'])
+    sc_instance = sc_class(transaction_data['address'])
+#    sc_instance = nusd1(transaction['specific_data']['address'])
+    funct = getattr(sc_instance, funct)
+    params_for_funct=transaction_data['params']
+    # adding singers address to the dict
+    params_for_funct['function_caller']=transaction_signer
+    try:
+        child_transactions_array = funct(params_for_funct)   ##TODO: modify the smart contract function behaviour to give this output
+    except Exception as e:
+        print('Exception durint smart contract function run', e)
+
+    sc_address = transaction_data['address']
+
+    for child_transaction in child_transactions_array:
+        child_transaction = child_transaction['transaction']
+        child_transaction_signatures = child_transaction['signatures']  # relevant for two-way-transfer, not otherwise
+        child_transaction_data = child_transaction['specific_data']
+        child_transaction_type = child_transaction['type']
+
+        if transaction_type == TRANSACTION_MINER_ADDITION:
+            print('Smart contracts can't add miners')
+            continue
+        if child_transaction_type==TRANSACTION_SC_INTERNAL:    #new type for SC's private data update
+            ###TBD
+            pass
+        if child_transaction_type==TRANSACTION_SMART_CONTRACT: #need to call recursively the same function
+            ###TBD
+            pass
+        if child_transaction_type == TRANSACTION_WALLET_CREATION:  # this is a wallet creation transaction
+            if child_transaction_data['custodian']!= sc_address:
+                print('Incorrect custodian for child transaction. Continuing with next child tx or exiting.')
+                continue
+            else:
+                add_wallet_pid(cur, child_transaction_data)   # may need to add a check here if SC is allowed to add wallets (ref: allowed_custodians)
```

Comment:
Thinking of a slightly modified way in which actual state update from transactions doesn't happen here but instead in the existing `update_state_from_transaction` only. 

This function will instead be 
`
def simplifyTransactions(transactions):
  simplified_transactions = []
  for transaction in transactions:
    if transaction.type == TRANSACTION_SMART_CONTRACT:
      child_transactions = simplifyTransactions([transaction])
      for child_transaction in child_transactions:
          if validate_sc_child_transaction(child_transaction):
            simplified_transactions.append(transaction)
    else:
      simplified_transactions.append(transaction)
`

This will take all the transactions in a block, simplify it recursively and make a final list of all transactions which can then be executed as before with minor changes to sender/custodian checks. 
@swapnilpawar31415 thoughts?

Reply:
Sounds good. There is only one minor complication. Two way transfer in which one party is an SC can get a bit convoluted. We can restrict SCs to only one-way transfer because two way transfer can anyway be simulated by simple call to SC with one side of the transfer being done by human sender and SC executing the second half itself or rejecting it and sending back what was received.